XenAdmin/SettingsPanels/NRPEEditPage.cs

```diff
@@ -0,0 +1,397 @@
+
+            foreach (CheckGroup checkGroup in CheckGroupList)
+            {
+                CheckDataGridView.Rows.Add(checkGroup.CheckThresholdRow);
+                CheckGroupDictByName.Add(checkGroup.Name, checkGroup);
+                CheckGroupDictByLabel.Add(checkGroup.NameCell.Value.ToString(), checkGroup);
+            }
+
+            UpdateOtherComponentBasedEnableNRPECheckBox(false);
+
+            AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlDark);
+            AllowHostsTextBox.GotFocus += AllowHostsTextBox_GotFocus;
+            AllowHostsTextBox.LostFocus += AllowHostsTextBox_LostFocus;
+
+            if (isHost)
+            {
+                PoolTipsPicture.Hide();
+                PoolTipsLabel.Hide();
+            }
+            InvalidParamToolTip = new ToolTip
+            {
+                IsBalloon = true,
+                ToolTipIcon = ToolTipIcon.Warning,
+                ToolTipTitle = Messages.INVALID_PARAMETER,
+                Tag = AllowHostsTextBox
+            };
+        }
+
+        public bool ValidToSave
+        {
+            get
+            {
+                InvalidParamToolTipText = "";
+                InvalidParamToolTip.ToolTipTitle = "";
+                CheckDataGridView.ShowCellToolTips = false;
+
+                if (!EnableNRPECheckBox.Checked)
+                {
+                    return true;
+                }
+
+                bool valid = true;
+                if (!IsAllowHostsValid())
+                {
+                    valid = false;
+                }
+
+                foreach (CheckGroup checkGroup in CheckGroupList)
+                {
+                    if (!checkGroup.CheckValue())
+                    {
+                        valid = false;
+                    }
+                }
+                return valid;
+            }
+        }
+
+        public bool HasChanged
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        public void Cleanup()
+        {
+        }
+
+        public void ShowLocalValidationMessages()
+        {
+            if (InvalidParamToolTip.Tag is Control ctrl)
+            {
+                InvalidParamToolTip.Hide(ctrl);
+                if (!InvalidParamToolTipText.Equals(""))
+                {
+                    HelpersGUI.ShowBalloonMessage(ctrl, InvalidParamToolTip, InvalidParamToolTipText);
+                }
+            }
+        }
+
+        public void HideLocalValidationMessages()
+        {
+            if (InvalidParamToolTip.Tag is Control ctrl && ctrl != null)
+            {
+                InvalidParamToolTip.Hide(ctrl);
+            }
+        }
+
+        public void SetXenObjects(IXenObject orig, IXenObject clone)
+        {
+            Clone = clone;
+            if (IsHost)
+            {
+                InitNRPEGeneralConfiguration();

```

Comment:
These two methods make server calls on the UI thread, which means that if the network is slow the UI will freeze. Please create a new AsyncAction in the XenModel/Actions/NRPE folder (see my commits in https://github.com/kc284/xenadmin/commits/private/bengangy/CP-44371) and move this code in there. This page should be running the async action, hook the Completed event and update the UI on success/failure accordingly, While the action is running, the UI should have some indication that something is happening on the background, so you may want to add a spinner panel saying something like "Retrieving NRPE configurarion..."